pkg/serial/handshake.go

```diff
@@ -77,212 +72,188 @@ func HandshakeClient(ctx context.Context, conn net.Conn, debug bool) error {
+			if err != nil {
+				log.Error(err)
+				return err
+			}
+
+			if n != len(rxbuf) {
+				return fmt.Errorf("packet size mismatch (expected %d, received %d)", len(rxbuf), n)
+			}
+
+			if bytes.Compare(rxbuf, txbuf) != 0 {
+				return fmt.Errorf("HandshakeClient: lossiness check FAILED")
+			}
+
+			// Tell the server we're good.
+			if _, err = conn.Write([]byte{flagAck}); err != nil {
+				return err
+			}
+
+			log.Infof("HandshakeClient: lossiness check PASSED")
 		}
-
-		log.Infof("client: lossiness check PASSED")
-
+		break
 	}
+	log.Debug("HandshakeClient: Connection established.")
+	conn.SetReadDeadline(time.Time{})
 
 	return nil
 }
 
-func readMultiple(conn net.Conn, b []byte) (int, error) {
-	if runtime.GOOS != "windows" {
-		return conn.Read(b)
-	}
-
-	// we want a blocking read, but the behaviour described in the remarks section
-	// here is a problem as we never get the syn in the same read as the rest.
-	// https://msdn.microsoft.com/en-us/library/aa363190(v=VS.85).aspx
-
-	// we know we're never reading single bytes for the handshake, so we hack this
-	// multi read path into place
-	// ideally we'd allow fine tweaking of the ReadIntervalTimeout so that we could
-	// toggle between this behaviour and blocking read at a windows level
-	n, err := conn.Read(b)
-	if err == nil && n == 1 {
-		cn, cerr := conn.Read(b[1:])
-		return cn + 1, cerr
-	}
-	return n, err
-}
-
 func HandshakeServer(ctx context.Context, conn net.Conn) error {
 	if tracing {
 		defer trace.End(trace.Begin(""))
 	}
-	syn := make([]byte, 3)
-	synack := make([]byte, 3)
-	buf := make([]byte, 2)
-	ack := make([]byte, 2)
 
-	// set the read deadline for timeout
-	// this has no effect on windows as the deadline is set at port open time
 	deadline, ok := ctx.Deadline()
 	if ok {
 		conn.SetReadDeadline(deadline)
 	}
 
-	log.Debug("server: reading syn")
-	// syn is 3 bytes as that will eventually syn us again if we're offset
-	if n, err := readMultiple(conn, syn); n != 2 || err != nil || syn[0] != flagSyn {
-		var msg string
+	buf1byte := make([]byte, 1)
+	syncBuf := make([]byte, 4096)
+
+	rand.Read(buf1byte)
+	pos := nextPos(buf1byte[0])
+
+	log.Debugf("HandshakeServer: Establishing connection...")
+	for {
+		// Sync packet is 2 bytes, however if we read more than 2
+		// it means buffer is not empty and data is not trusted for this sync.
+
+		n, err := io.ReadAtLeast(conn, syncBuf, 2)
 		if err != nil {
-			msg = fmt.Sprintf("server: failed to read expected SYN: n=%d, err=%s", n, err)
-		} else if syn[0] != flagSyn {
-			msg = fmt.Sprintf("server: did not receive SYN (read %v bytes): %#x != %#x", n, flagSyn, syn[0])
-			conn.Write([]byte{flagNak})
-		} else {
-			msg = fmt.Sprintf("server: received SYN (read %d) bytes", n)
+			return err
 		}
 
-		// to aid in debug we always dump the full handhsake
-		log.Debug(msg)
-		log.Debugf("server: read %v bytes: ", n)
-		for i := 0; i < n; i++ {
-			log.Debugf("%#x ", syn[i])
+		if n > 2 {
```

Comment:
Try assuming that `syncBuf[n-2:]` is the most _recent_ syn packet and continue with that. It works if there are only two bytes read as well as if there are 20 queued packets.